File: leetcode/c++/LCR_150_cong_shang_dao_xia_da_yin_er_cha_shu_ii_lcof.cpp
// https://leetcode.cn/problems/cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> decorateRecord(TreeNode* root) {
        vector<vector<int>> res;
        queue<TreeNode*> que;
        if (root == nullptr)
            return res;
        que.push(root);
        while (!que.empty()) {
           int i = que.size();
           vector<int> array;
           while (i--) {
                TreeNode* tmp = que.front();
                que.pop();
                if (tmp->left != nullptr)
                    que.push(tmp->left);
                if (tmp->right != nullptr)
                    que.push(tmp->right);
                array.push_back(tmp->val);
           }
           res.push_back(array);
        }
        return res;
    }
};

// 时间复杂度为O(n)，空间复杂度为O(n)

File: leetcode/golang/LCR_150_cong_shang_dao_xia_da_yin_er_cha_shu_ii_lcof.go
type TreeNode struct {
         Val int
         Left *TreeNode
         Right *TreeNode
}

func decorateRecord(root *TreeNode) [][]int {
    var res [][]int
    var que []*TreeNode
    if root != nil {
        que = append(que, root)
    } else {
        return res
    }
    for len(que) != 0 {
        var tmp []int
        for i := len(que); i > 0; i-- {
            node := que[0]
            que = que[1:]
            tmp = append(tmp, node.Val)
            if node.Left != nil {
                que = append(que, node.Left)
            }
            if node.Right != nil {
                que = append(que, node.Right)
            }
        }
        res = append(res, tmp)
    }
    return res
}